File: tests/versioning/surface/arms.py
"""Reading the arms of a Rust match block: where each arm begins, and the
literals in its pattern. This is how the command vocabulary is recovered from
the dispatcher itself rather than from a list somebody has to remember to
update."""

from __future__ import annotations

from .source import NOT_FOUND, ONE, TWO, ZERO, Source, SurfaceError

OPENERS = {"{": "}", "(": ")", "[": "]"}
CLOSERS = {"}": "{", ")": "(", "]": "["}


def match_arm_patterns(source: Source, start: int, stop: int) -> list:
    """Literals in the pattern of every arm of the match block spanning start..stop.

    Splits arms by scanning for `=>` at the block's own nesting depth, then takes
    only the literals that sit at that same depth inside the pattern. A literal
    in an arm *body* is nested inside the body's braces or parens, so it cannot
    be mistaken for a pattern.
    """
    mask = source.mask
    patterns = []
    depth = ZERO
    depths = []
    boundary = start
    index = start
    while index < stop:
        char = mask[index]
        if char in OPENERS:
            depth += ONE
            index += ONE
            continue
        if char in CLOSERS:
            depth -= ONE
            index += ONE
            continue
        if depth == ZERO and mask.startswith("=>", index):
            patterns.append((boundary, index))
            body = index + TWO
            while body < stop and mask[body] == " ":
                body += ONE
            if body < stop and mask[body] == "{":
                body = source.balanced_end(body, "{", "}")
                while body < stop and mask[body] in " ,\n":
                    body += ONE
                boundary = body
                index = body
                continue
            inner = ZERO
            while body < stop:
                token = mask[body]
                if token in OPENERS:
                    inner += ONE
                elif token in CLOSERS:
                    inner -= ONE
                elif token == "," and inner == ZERO:
                    break
                body += ONE
            boundary = body + ONE
            index = boundary
            continue
        index += ONE
    del depths
    names = []
    for begin, end in patterns:
        for offset, _stop, value in source.literals:
            if begin <= offset < end and pattern_depth(mask, begin, offset) == ZERO:
                names.append(value)
    return names


def pattern_depth(mask: str, begin: int, offset: int) -> int:
    depth = ZERO
    for index in range(begin, offset):
        char = mask[index]
        if char in OPENERS:
            depth += ONE
        elif char in CLOSERS:
            depth -= ONE
    return depth

File: tools/src/surface/arms.rs
//! Reading the arms of a Rust match block: where each arm begins, and the
//! literals in its pattern. This is how the command vocabulary is recovered
//! from the dispatcher itself rather than from a list somebody has to
//! remember to update.

use super::source::Source;

fn opens(byte: u8) -> bool {
    matches!(byte, b'{' | b'(' | b'[')
}

fn closes(byte: u8) -> bool {
    matches!(byte, b'}' | b')' | b']')
}

/// Literals in the pattern of every arm of the match block spanning
/// `start..stop`.
///
/// Arms are split at `=>` found at the block's own nesting depth, and only
/// literals at that same depth inside a pattern are taken. A literal in an arm
/// body is nested inside the body's braces or parentheses, so it cannot be
/// mistaken for a pattern.
pub(crate) fn match_arm_patterns(
    source: &Source,
    start: usize,
    stop: usize,
) -> Result<Vec<String>, String> {
    let mask = source.mask.as_bytes();
    let mut patterns = Vec::new();
    let mut depth = 0isize;
    let mut boundary = start;
    let mut index = start;
    while index < stop {
        let byte = mask[index];
        if opens(byte) {
            depth += 1;
            index += 1;
            continue;
        }
        if closes(byte) {
            depth -= 1;
            index += 1;
            continue;
        }
        if depth == 0 && mask[index..].starts_with(b"=>") {
            patterns.push((boundary, index));
            let mut body = index + 2;
            while body < stop && mask[body] == b' ' {
                body += 1;
            }
            if body < stop && mask[body] == b'{' {
                body = source.balanced_end(body, b'{', b'}')?;
                while body < stop && matches!(mask[body], b' ' | b',' | b'\n') {
                    body += 1;
                }
                boundary = body;
                index = body;
                continue;
            }
            let mut inner = 0isize;
            while body < stop {
                let token = mask[body];
                if opens(token) {
                    inner += 1;
                } else if closes(token) {
                    inner -= 1;
                } else if token == b',' && inner == 0 {
                    break;
                }
                body += 1;
            }
            boundary = body + 1;
            index = boundary;
            continue;
        }
        index += 1;
    }
    let mut names = Vec::new();
    for (begin, end) in patterns {
        for (offset, _, value) in &source.literals {
            if (begin..end).contains(offset) && pattern_depth(mask, begin, *offset) == 0 {
                names.push(value.clone());
            }
        }
    }
    Ok(names)
}

fn pattern_depth(mask: &[u8], begin: usize, offset: usize) -> isize {
    mask[begin..offset].iter().fold(0, |depth, &byte| {
        if opens(byte) {
            depth + 1
        } else if closes(byte) {
            depth - 1
        } else {
            depth
        }
    })
}
